### PR TITLE
OCD-4203 - Implement basic Cognito authentication/authorization

### PR DESCRIPTION
**Pull request by tmy1313**

### CODE REVIEW

File: chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/UserManagementController.java [hunk 1]
```diff
@@ -347,7 +357,11 @@ public DeletedUser deleteUser(@PathVariable("userId") Long userId)
     public @ResponseBody User getUser(@PathVariable("id") Long id)
             throws UserRetrievalException {
 
-        return userManager.getUserInfo(id);
+        if (ff4j.check(FeatureList.SSO)) {
+            return cognitoAuthenticationManager.getUserInfo("tmy1313@gmail.com");
```
Comment (kekey1) [inline chpl/chpl-api/src/main/java/gov/healthit/chpl/web/controller/UserManagementController.java]: Do you want yourself hardcoded in here?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/FeatureList.java [hunk 2]
```diff
@@ -9,4 +9,5 @@ private FeatureList() {
     public static final String ERD_PHASE_2_GRACE_PERIOD_END = "erd-phase-2-grace-period-end";
     public static final String ERD_PHASE_3 = "erd-phase-3";
     public static final String SUBSCRIPTIONS = "subscriptions";
+    public static final String SSO = "sso";
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/FeatureList.java:12]: Make sure to add this to all environments (I happened to look at prod and did not see it there)

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/auth/authentication/CognitoJwtUserConverter.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        user.setAuthenticated(true);
+        user.setId(decodeJwt.getClaim("id").asLong());
+        user.setFullName(decodeJwt.getClaim("name").asString());
+        user.setSubjectName(decodeJwt.getClaim("email").asString());
+
+        decodeJwt.getClaim("cognito:groups").asList(String.class).stream()
+                .forEach(role -> user.addPermission(new GrantedPermission(role)));
+
+        //decodeJwt.getClaims().entrySet().stream()
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/auth/authentication/CognitoJwtUserConverter.java]: keep? delete?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/auth/jwt/CognitoRsaKeyProvider.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import com.auth0.jwk.JwkProvider;
+import com.auth0.jwk.JwkProviderBuilder;
+import com.auth0.jwt.interfaces.RSAKeyProvider;
+
+public class CognitoRsaKeyProvider implements RSAKeyProvider {
+    private final URL awsKidStoreUrl;
+    private final JwkProvider provider;
+
+    public CognitoRsaKeyProvider(String awsCognitoRegion, String awsUserPoolId) {
+        String url = String.format("https://cognito-idp.%s.amazonaws.com/%s/.well-known/jwks.json", awsCognitoRegion, awsUserPoolId);
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/auth/jwt/CognitoRsaKeyProvider.java]: It feels like some of this should be in an environment property or at least a constant

**Review by kekey1 — CHANGES_REQUESTED** [review]

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/auth/authentication/CognitoJwtUserConverter.java [hunk 5]
```diff
@@ -23,13 +23,14 @@ public class CognitoJwtUserConverter {
     private String region;
     private String userPoolId;
     private String clientId;
-
+    private String tokenizezRsaKeyUrl;
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/auth/authentication/CognitoJwtUserConverter.java]: Is this the right word here or is it a typo? Genuine question. I feel like there is an extra "z" or is it spelled that way on purpose like we might do with "clazz" ?

**Review by kekey1 — DISMISSED** [review]

**Review by andlar — DISMISSED** [review]

**Review by andlar — DISMISSED** [review]

File: chpl/chpl-resources/src/main/resources/environment.properties [hunk 6]
```diff
@@ -311,13 +311,25 @@ spring.redis.threads=32
...
+cognito.clientId=SECRET
+cognito.tokenizezRsaKeyUrl=https://cognito-idp.%s.amazonaws.com/%s/.well-known/jwks.json
+##############################################
+
 ######  Dynamic Scheduling of Download Files     ######
 download2011Hour=10
 download2011Minute=35
 download2014Hour=12
 download2014Minute=45
+download2015Hour=15
```
Comment (kekey1) [inline chpl/chpl-resources/src/main/resources/environment.properties]: I don't think we need this? We don't create a 2015 file anymore

**Review by kekey1 — CHANGES_REQUESTED** [review]

**Review by kekey1 — APPROVED** [review]

**Review by andlar — APPROVED** [review]

**Pull request merged**